src/OrbitBase/include/OrbitBase/WhenAny.h

```diff
@@ -0,0 +1,107 @@
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_WHEN_ANY_H_
+#define ORBIT_BASE_WHEN_ANY_H_
+
+#include <absl/base/thread_annotations.h>
+#include <absl/synchronization/mutex.h>
+
+#include <variant>
+
+#include "OrbitBase/Future.h"
+#include "OrbitBase/FutureHelpers.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Promise.h"
+
+namespace orbit_base_internal {
+
+// A small helper type trait that maps `void` to `std::monostate` and every other `T` to itself.
+template <typename T>
+struct VoidToMonostate {
+  using type = T;
+};
+
+template <>
+struct VoidToMonostate<void> {
+  using type = std::monostate;
+};
+
+template <typename T>
+using VoidToMonostate_t = typename VoidToMonostate<T>::type;
+
+template <typename... Ts>
+class SharedStateWhenAny {
+  orbit_base::Promise<std::variant<Ts...>> promise ABSL_GUARDED_BY(mutex);
+  mutable absl::Mutex mutex;
+
+ public:
+  template <size_t index>
+  void SetResult(const typename std::variant_alternative_t<index, std::variant<Ts...>>& argument)
+      ABSL_LOCKS_EXCLUDED(mutex) {
+    absl::MutexLock lock{&mutex};
+    if (promise.HasResult()) return;
+
+    promise.SetResult(std::variant<Ts...>{std::in_place_index_t<index>{}, argument});
+  }
+
+  [[nodiscard]] orbit_base::Future<std::variant<Ts...>> GetFuture() const {
+    absl::MutexLock lock{&mutex};
+    return promise.GetFuture();
+  }
+};
+
+// A small helper type trait that checks whether a certain index of a parameter pack `Args` is of
+// type `std::monostate`. If yes the resulting type will be `std::true_type`, otherwise
+// `std::false_type`.
+template <size_t index, typename... Args>
+using IsMonostate =
+    std::is_same<std::variant_alternative_t<index, std::variant<VoidToMonostate_t<Args>...>>,
+                 std::monostate>;
+
+template <typename... Args, std::size_t... Indexes>
+orbit_base::Future<std::variant<VoidToMonostate_t<Args>...>> WhenAnyImpl(
+    orbit_base::Future<Args>... futures, std::index_sequence<Indexes...> /* indexes */) {
+  ORBIT_CHECK(futures.IsValid() && ...);
+
+  auto shared_state =
+      std::make_shared<orbit_base_internal::SharedStateWhenAny<VoidToMonostate_t<Args>...>>();
+
+  (RegisterContinuationOrCallDirectly(
+       futures,
+       // Having the lambda expression taking a parameter pack as its arguments allows us to express
+       // 0 arguments (void) and 1 argument (any other T) in a single continuation.
+       [shared_state](const auto&... arguments) {
+         if constexpr (IsMonostate<Indexes, Args...>::value) {
+           shared_state->template SetResult<Indexes>(std::monostate{});
+         } else {
+           // `arguments` is a parameter pack of at most 1 argument. In this if-constexpr-branch we
+           // already know it's exactly one argument, so `(arguments, ...)` just unfolds to a single
+           // argument.
+           shared_state->template SetResult<Indexes>((arguments, ...));
+         }
+       }),
+   ...);
```

Comment:
I have to say, I have no idea what `...` is doing here.

Reply:
That's where all the magic happens. :sweat_smile: 

You can see that as a unrolled for-loop over all the indices in `Indices`. For each loop iteration `Indices` will be replaced by the current index.